PEP 785: New methods for easier handling of `ExceptionGroup`s (#4357)
python/peps: peps/pep-0790.rst

        first, *rest = group.flat_exceptions()  # get the whole traceback :-)
        if not rest:
            raise first
        raise

However, this innocent-seeming code has a problem: ``raise first`` will do
``first.__context__ = group`` as a side effect.  This discards the original
context of the error, which may contain crucial information to understand why
the exception was raised.  In many production apps it also causes tracebacks
to balloon from hundreds of lines, to tens or even `hundreds of thousands of
lines <https://github.com/python-trio/trio/issues/2001#issuecomment-931928509>`__ -
a volume which makes understanding errors far more difficult than it should be.

A new ``BaseException.preserve_context()`` method would be a discoverable,
readable, and easy-to-use solution for these cases.

Specification
=============

Adding ``BaseExceptionGroup.flat_exceptions()``
-----------------------------------------------

A new method ``flat_exceptions()`` will be added to ``BaseExceptionGroup``, with the
following signature:

.. code-block:: python

    def flat_exceptions(self) -> list[BaseException]:
        """
        Return a flat list of all 'leaf' exceptions.

>         If the exception is not a ``BaseExceptionGroup``, return ``[eg]``.

Reviewer: ```suggestion
        If the exception is not a ``BaseExceptionGroup``, return ``[self]``.
```
"eg" is not defined. I assume it stands for ExceptionGroup, but we've just established that it isn't one :)

Author: In fact this docstring is just out of date; since this method is defined on `BaseExceptionGroup` (not `BaseException`), `self` _must_ be a group.  Will update.

following signature:

.. code-block:: python

    def preserve_context(self) -> contextlib.AbstractContextManager[Self]:
        """
        Context manager that preserves the exception's __context__ and
        __suppress_context__ attributes.

        When entering the context, the current values of __context__ and
        __suppress_context__ are saved. When exiting, these saved values are
        restored to the exception. This allows raising an exception inside
        an except block without changing its context chain.
        """

Usage example:

.. code-block:: python

    # We're an async web framework, where user code can raise an HTTPException
    # to return a particular HTTP error code to the client.  However, it may
    # (or may not) be raised inside a TaskGroup, so we need to use `except*`;
    # and if there are *multiple* such exceptions we'll treat that as a bug.
    try:
        user_code_here()
    except* HTTPException as group:
        first, *rest = group.flat_exceptions()
        if rest:
            raise  # handled by internal-server-error middleware
        ...  # logging, cache updates, etc.
        with first.preserve_context():

>             raise first

Reviewer: Maybe more a question for Discuss, but: If the usage is always going to be `with e.preserve_context(): raise e`, does it make more sense to make a single method that does both of these things? For example, a method `e.raise_with_preserved_context()`.

Without ``.preserve_context()``, this could would have to either:

* arrange for the exception to be raised *after* the ``except*`` block,
  making code difficult to follow in nontrivial cases, or
* discard the existing ``__context__`` of the ``first`` exception, replacing
  it with an ``ExceptionGroup`` which is simply an implementation detail, or

> * use ``try/except`` instead of ``except*``, handling  the possiblity that

Reviewer: ```suggestion
* use ``try/except`` instead of ``except*``, handling the possibility that
```

re-raising a leaf exception inside an ``except*`` block would override any
existing context.

Rejected Ideas
==============

Add utility functions instead of methods
----------------------------------------

Rather than adding methods to exceptions, we could provide utility functions:

.. code-block:: python

    def flat_exceptions(exc):
        if not isinstance(exc, BaseExceptionGroup):
            return [exc]
        # implementation

There are however several reasons to prefer methods: there's no obvious place
where helper functions should live, they take exactly one argument which must
be an instance of ``BaseException``, and methods are both more convenient and
more discoverable.

Add ``BaseException.as_group()`` (or group methods)
---------------------------------------------------

Our survey of ``ExceptionGroup``-related error handling code also observed
many cases of duplicated logic to handle both a bare exception, and the same
kind of exception inside a group (often incorrectly, motivating
``.flat_exceptions()``).

We briefly `proposed <https://github.com/python/cpython/issues/125825>`__
adding ``.split(...)`` and ``.subgroup(...)`` methods too all exceptions,

> before considering ``.flat_exceptions()`` made me feel this was too clumsy.

Reviewer: The sentences moves from "we" to "me".

As a cleaner alternative, we sketched out an ``.as_group()`` method:

.. code-block:: python

    def as_group(self):
        if not isinstance(self, BaseExceptionGroup):
            return BaseExceptionGroup("", [self])
        return self

However, applying this method to refactor existing code was a negligible
improvement over writing the trivial inline version.  We also hope that many
current uses for such a method will be addressed by ``except*`` as older
Python versions reach end-of-life.

We recommend documenting a "convert to group" recipe for de-duplicated error
handling, instead of adding group-related methods to ``BaseException``.

Footnotes
=========

> .. [#numbers]

Reviewer 2: This isn't referenced in the text and breaks the build:
```
/home/runner/work/peps/peps/peps/pep-0790.rst:350: WARNING: Footnote [#] is not referenced. [ref.footnote]
```
https://github.com/python/peps/actions/runs/14349585387/job/40225640358?pr=4357#step:5:136

      `two <https://github.com/mhdzumair/MediaFusion/blob/ff906378f32fb8419ef06c6f1610c08946dfaeee/scrapers/base_scraper.py#L375-L386>`__,
      `three <https://github.com/SonySemiconductorSolutions/local-console/blob/51f5af806336e169d3dd9b9f8094a29618189f5e/local-console/src/local_console/commands/server.py#L61-L67>`__,
      `four <https://github.com/SonySemiconductorSolutions/local-console/blob/51f5af806336e169d3dd9b9f8094a29618189f5e/local-console/src/local_console/commands/broker.py#L66-L69>`__,
      `five <https://github.com/HexHive/Tango/blob/5c8472d1679068daf0f041dbbda21e05281b10a3/tango/fuzzer.py#L143-L160>`__,
      `six <https://github.com/PaLora16/ExceptionsGroupsValidators/blob/41152a86eec695168fdec74653694658ddc788fc/main.py#L39-L44>`__,
      `seven <https://github.com/reactive-python/reactpy/blob/178fc05de7756f7402ed2ee1e990af0bdad42d9e/src/reactpy/backend/starlette.py#L164-L170>`__)

    indicating that more than a quarter of _all_ hits for this fairly general
    search would benefit from the methods proposed in this PEP.

> .. [#catch-raw-group]

Reviewer 2: Likewise:
```
/home/runner/work/peps/peps/peps/pep-0790.rst:387: WARNING: Footnote [#] is not referenced. [ref.footnote]
```

PEP: 790

> Title: New methods for easier handling of ExceptionGroups

Reviewer 3: ```suggestion
Title: New methods for easier handling of :class:`ExceptionGroup`\ s
```
